### DISCUSSION TITLE
Dampening

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#28] [Ideas]
It strikes me as unworkable that forecast dampening be restricted to certain hours of the day.

In the height of Summer the curve looks way different to the depths of Winter.

Should dampening be based on sunrise/sunset times, and use a different algorithm than simply hours of the day? Possibly even account for shading coming from different directions at different times of year 🤷‍♂️.

**Comment by Nilogax** [comment]
I have built a mechanism that does a daily update of the dampening settings based on an hourly comparison of actual generation vs forecast over several previous days. That seems to take account of the changes in shading for my array, but I'm no HA/Python expert so have probably gone about it in a very sub optimal way!

**Reply by Noodleyman** [reply]
@Nilogax , would you be willing to share your code? I'm working on the exact same thing at the moment and have it building the data  which I want to send to the damperning service, however home assistant sees the string of values as a "list", thus it refuses to pass it properly in the service call and I'm struggling to figure out how to get it to convert from list to string.

**Reply by gcoan** [reply]
It might be an interesting future enhancement to the solcast integration to include this

... but, I'm not sure that dampening is always and entirely the correct thing to do.  I have a lot of solar on my house (44 panels in total in 6 strings across 3 arrays) and other than one string on the garage roof which gets shaded by next door in the winter, I don't have any shading at all.

But in the summer I never get the solcast forecast generation.
Why, because its a forecast, and the 'core' 50% forecast is just that, its the forecast that is predicted to happen 50% of the time.  I personally look at the 10% forecast as well (1 in 10 worst case) to give me a range that the solar generation will be in and my forecast is always between the two, usually somewhere around 30% of the way between the 10% and the 50%.

I use a weighted figure of the 10% to the 50% forecast in predbat for my solar battery planning.
e.g. today:

![image](https://github.com/BJReplay/ha-solcast-solar/assets/142018870/c9f90428-03da-4199-a6e5-2723a373d1ac)

**Reply by Nilogax** [reply]
That's a fair chunk of generation there @gcoan! As they saying goes, "All models are wrong but some are useful", so there's always likely to be some divergence between any forecast and reality.  If there are known, predictable local factors that will tend to increase that divergence then I think that including them pushes the revised forecast a little more towards the "useful" side.  Perhaps with a larger array the impact of shading is smaller and can be accomodated as you suggested, but when one of my two strings can be severely shaded at certain times of day/year it's (IMO) worth including that in the forecast so predbat has the best input data to work from.   

@Noodleyman my solution is built using some automations to log the generation each hour, then I used pyscript (https://github.com/custom-components/pyscript) to do the calculations and call the services.  That means I can just build the new dampening settings as a string and submit that via the service call.   I was learning HA & Python as I went along so it's a bit of a mess but as long as i works...

**Reply by Nilogax** [reply]
@Noodleyman if you're still working with dampening you may have noticed that in recent versions the Solcast data wasn't particularly responsive to changes in dampening settings. I had raised an issue about this and have just updated it with a potential fix/workaround that I found today - see https://github.com/BJReplay/ha-solcast-solar/issues/44

**Reply by autoSteve** [reply]
@Noodleyman, moved to discussion. https://github.com/BJReplay/ha-solcast-solar/discussions/54.

I think a reversion may be in order, but happy to hear the opinions of others.

**Reply by Noodleyman** [reply]
Morning,

Thanks for sharing the additional info. Can confirm I've run into some issues with the Solcast dampening results, but not quite as described in #44 and #54. 

On at least 3 occasions since I have set it up, Solcast seems to have crashed (seems but has not). ALL data points get set to 0. calling the API for an update does not reset it even though it is successful and even removing the Dampening settings doesn't reset it. The only way to get a forecast back that I have found is to restart HA.

Ive only been able to trigger the issue when Dampening is configured, and I was able to trigger it once simply by setting dampening configuration. 

I'm going to back out the use of Dampening for the time being, as when the issue happens the knock on impact to battery charging etc is not ideal. 

**Example:**
Happened at 00:03 (not a scheduled Forecast update time,  01:30 is the next update), but at the time the dampening values were updated via automation
![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/d1168170-4b67-443f-8a9c-769dfb3f90e0)
![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/d9daa3bf-b1d4-4fe7-a3d3-97af4daaef9c)
![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/c0c00dd9-2fd7-49bf-bf93-0874777442f8)
![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/fce3c29b-2261-4c95-b3b5-62330d4aea72)

All sensorts reported 0 (but now have a value as I restarted already today), example:
![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/6a5788c9-7fa4-40e4-8be8-75f0e304644b)

My dampening values:
![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/4660021b-1a16-460f-8d47-8316c7432137)

at 00:03, those dampening values were updated by automation, which caused everything to fail. the dampening values are derrived values based on the time of day the sun comes up, thus update once per night

**Reply by Nilogax** [reply]
@Noodleyman - just noticed that @autoSteve has released 4.0.33 which includes the code reversion mentioned in the other discussion. I'd suggest downloading that to see if it has any impact.

**Reply by Nilogax** [reply]
@Noodleyman - see https://github.com/BJReplay/ha-solcast-solar/discussions/60 for the 0.33 pre release info.

**Reply by Noodleyman** [reply]
ohhh, amazing. thank you! upgrading to that release now to see what happens.

**Reply by autoSteve** [reply]
More dampening goodness. https://github.com/BJReplay/ha-solcast-solar/discussions/63.
